plasmapy/diagnostics/langmuir.py

```diff
@@ -1,153 +1,1073 @@
-"""
-Defines the Langmuir analysis module as part of the diagnostics package.
-"""
+"""Defines the Langmuir analysis module as part of the diagnostics package."""
 
-import numpy as np
+import plasmapy.constants as const
 import astropy.units as u
+import copy
+
+from plasmapy import utils
+
+from astropy.visualization import quantity_support
+import numpy as np
+import matplotlib.pyplot as plt
+
 
+def fit_func_lin(x, x0, y0, c0):
+    r"""Linear fitting function."""
+    return y0 + c0 * (x - x0)
 
-class Probe():
-    r"""
-    Class containing various Langmuir probe properties.
 
-    For the moment cylindrical probes are assumed as spherical ones are
-    generally impractical to make.
+def fit_func_double_lin(x, x0, y0, c0, c1):
+    r"""Piecewise linear fitting function. (x0, y0) denotes the location of
+    the knee of the transition, with c0 and c1 being the left and right-hand
+    slopes, respectively.
+    """
+    return np.piecewise(x, x < x0, [lambda x: y0 + c0 * (x - x0),
+                                    lambda x: y0 + c1 * (x - x0)])
+
+
+class Characteristic:
+    r"""Class representing a I-V probe characteristic for convenient access.
+    Supports units.
 
     Attributes
     ----------
-    area : ~astropy.units.Quantity or ~numpy.ndarray
-        Area or array of areas of the probe(s) in units convertible to m^2.
+    bias : Quantity, ndarray
+        Array of applied probe biases in units convertible to V.
 
-    config : str, optional
-        Electrode configuration. Possible options are `'single'`
-        (default), `'double'`, `'double-asym'`, `'triple'`, `'tetra'`,
-        `'penta'`. The electrode configuration must correspond to the
-        number of areas provided.
+    current : Quantity, ndarray
+        Array of applied probe currents in units convertible to A.
 
     """
-    def __init__(self, area=u.m**2, config='single'):
-        # Initialize probe properties
-        self.area = area
-        self.config = config
 
+    def __init__(self, bias, current):
```

Comment:
Did you try the `@check_quantity` decorator? I see you made a `check_validity` method - does that mean you did, and it didn't work for class methods? That's something I'll have to test anyway...

Reply:
Yes, but I'd like to consolidate all checks into a function that can also be called in the code during execution. Someone might initialize a Characteristic that passes the checks but mess with the values of the array before passing it as a parameter. I also need to extent `check_validity` beyond what a decorator would be capable of, I think. For example ensuring the size of the arrays match.